[CALCITE-2446] Fix for lateral join in sql views

ptrbojko opened the pull request.

vlsi commented on plus/src/main/java/org/apache/calcite/chinook/CodesFunction.java:
```diff
@@ -0,0 +1,67 @@
+        return Linq4j.asEnumerable(new String[][]{
+            new String[]{"HASHCODE", "" + name.hashCode()},
+            new String[]{"BASE64",
+                Base64.encodeBase64String(name.getBytes(Charset.defaultCharset()))}
```
Please refrain from using `defaultCharset` in tests

vlsi commented on plus/src/main/java/org/apache/calcite/chinook/CodesFunction.java:
```diff
@@ -0,0 +1,67 @@
+                                                       SchemaPlus schema,
+                                                       String tableName) {
+        if (name == null) {
+          return Linq4j.asEnumerable(new String[][]{}).asQueryable();
```
Should this be `emptyEnumerable` ?

vlsi commented on plus/pom.xml:
```diff
@@ -55,6 +55,11 @@ limitations under the License.
     </dependency>
+    <dependency>
+      <groupId>commons-codec</groupId>
+      <artifactId>commons-codec</artifactId>
```
It does look more like a `<scope>test</scope>` dependency, doesn't it?
I would like to refrain from adding regular dependencies for test purposes.

vlsi commented on core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:
```diff
@@ -2049,7 +2049,7 @@ private SqlNode registerFrom(
               forceNullable,
               lateral);
-      if (newExpr != expr) {
+      if (newExpr != expr) { // && (isLateral(expr) && !isLateral(newExpr))
```
I'm afraid this comment does not clarify much.

ptrbojko commented:
@vlsi I've patched items from your review.

zinking commented:
you may also want to squeeze the fix into 1 commit.

vlsi commented:
@ptrbojko , I have no clue if `validator.setIdentifierExpansion(true);` is a proper approach or not.

ptrbojko commented:
I've squashed the commits.
I've removed unneeded (as for the fix sake) - validator.setIdentifierExpansion(true); . There are view expansion methods in calcite and this change was a little step to make them look similar. But this wont fix the bug.

vlsi commented on core/src/test/java/org/apache/calcite/util/Smalls.java:
```diff
@@ -126,6 +126,29 @@ public RelDataType getRowType(RelDataTypeFactory typeFactory) {
     };
   }
 
+  private static QueryableTable simpleTable(String s) {
```
Is `simpleTable` vital for the test?
Can you reuse existing table function?

ptrbojko commented:
@vlsi  I've reused existing tbl function instead of creating new one. As desired.
